Review

rmnkubyshkin opened the pull request:
PR review to main

rmnkubyshkin committed:
feat: Implemented SideForm

rmnkubyshkin committed:
fix: Deleted comments

gennady-bars commented on src/index.tsx:
```diff
@@ -1,31 +1,59 @@
+};
 
 const App = () => {
+	const [open, setOpen] = useState(false);
```
* `[open, setOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете именно **меню**

gennady-bars commented on src/index.tsx:
```diff
@@ -1,31 +1,59 @@
 
 const App = () => {
+	const [open, setOpen] = useState(false);
+	const [option, setOption] = useState(defaultStyle);
+
+	useEffect(() => {
+	}, [open,handleOptionClick])
+
+	function handleOptionClick(option: FormStyleProps, hook: (opt: FormStyleProps) => void) {
+		hook(option);
+		setOpen(!open);
+	}
```
* Вся логика формы должна быть описана внутри `ArticleParamsForm`, чтобы не засорять код `App` и пропсы `ArticleParamsForm`

gennady-bars commented on src/index.tsx:
```diff
@@ -1,31 +1,59 @@
+
 	return (
 		<main
 			className={clsx(styles.main)}
```
**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается

gennady-bars commented on src/components/side-form/SideForm.tsx:
```diff
@@ -0,0 +1,128 @@
+			<aside className={styles.container} style={{transform: translate}}>
+				<form className={styles.form}>
+					<div className={styles.title}>
+						<h1 className={styles.h1} >Задайте параметры</h1>
```
* Вместо обычного тега `h1` нужно использовать компонент `Text`, который дается Вам изначально. `Text` нужно превратить по семантике в тег `h2` через пропсы: `as=("h2")`

gennady-bars commented on src/components/side-form/SideForm.tsx:
```diff
@@ -0,0 +1,128 @@
+								onClick={resetStyle}
+						/>
+						<Button title='Применить' htmlType='submit' type='apply'
+								onClick={submitStyle}/>
```
`onClick={submitStyle}`

Это нужно удалить с кнопки

**Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только.

gennady-bars commented on src/components/side-form/SideForm.tsx:
```diff
@@ -0,0 +1,128 @@
+
+	const sepRef = useRef(null);
+
+	const [translate, setTranslate] = useState('translate(-616px)');
```
`	const [translate, setTranslate] = useState('translate(-616px)');
`

Вам это не требуется. В проекте даются стили плавного открытия и закрытия формы. В стилях формы есть `container_open`. Там как раз `transform: translate(0);`

gennady-bars commented on src/components/side-form/SideForm.tsx:
```diff
@@ -0,0 +1,128 @@
+
+	const [translate, setTranslate] = useState('translate(-616px)');
+	useEffect(() => {
+		isOpen? setTranslate('translate(0)'):setTranslate('translate(-616px)')
```
* По заданию нужно использовать библиотеку `clsx` для объединения 2х и более стилей

Там будет добавляться `container_open` и удаляться в зависимости от `isOpen`

gennady-bars replied:
**Можно лучше**

* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* Хорошо реализовали форму из предоставленных компонентов
* Отлично, что нет типов `any`
* Отлично типизировали проект


но есть некоторые недочеты:

* `[open, setOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете именно **меню**
* Меню настроек должно закрываться при клике **снаружи** блока. Нужно использовать хук для этого. Примеры хуков есть в проекте в папке `select`
* Вся логика формы должна быть описана внутри `ArticleParamsForm`, чтобы не засорять код `App` и пропсы `ArticleParamsForm`
* Вместо обычного тега `h1` нужно использовать компонент `Text`, который дается Вам изначально. `Text` нужно превратить по семантике в тег `h2` через пропсы: `as=("h2")`
* **Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только.  
* По заданию нужно использовать библиотеку `clsx` для объединения 2х и более стилей


**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается
* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

rmnkubyshkin committed:
fix: Added new fix from first pull request

	1. Added mouse handler by clicking ouside form
	2. Used clxs for handling styles
	3. Created App.tsx file for App class
	4. Used Text components instead of h1 tag
	5. Renamed state to 'isOpenMenu' for more understanding code

gennady-bars commented on src/App.tsx:
```diff
@@ -0,0 +1,31 @@
+
+export const App = () => {
+	const [styleArticle, setStyleArticle] = useState(defaultStyle);
+	const [isMenuOpen, setIsMenuOpen] = useState(false);
```
`const [isMenuOpen, setIsMenuOpen] = useState(false);` нужно изначально внутри `ArticleParamsForm` объявить. Это только внутренняя логика формы. `App` не должен о ней знать

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,49 @@
+	const formRef: React.RefObject<HTMLDivElement> = useRef<HTMLDivElement>(null);
 
-export const ArticleParamsForm = () => {
+	useOutsideClickClose({
+		isOpen,
+		rootRef: formRef,
+		onChange: () => {
+			onClick?.()
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для правильного навешивания и удаления обработчиков `Escape` и оверлея (клика снаружи блока):

```js
import { useEffect } from "react";

type TUseClose = {
  isOpen: boolean;
  onClose: () => void;
  rootRef: React.RefObject<HTMLElement>;
};
// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export function useClose({isOpen, onClose, rootRef}: TUseClose) {
  useEffect(() => {
    if (!isOpen) return; // останавливаем действие эффекта, если закрыто

   
    function handleClickOutside(event: MouseEvent) {
      const { target } = event;
      const isOutsideClick =
        target instanceof Node && // проверяем, что это `DOM`-элемент
        rootRef.current &&
        !rootRef.current.contains(target);  // проверяем, что кликнули на элемент, который находится не внутри нашего блока
      if (isOutsideClick) {
        onClose();
      }
    }

    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    document.addEventListener("mousedown", handleClickOutside);

    //  обязательно удаляем обработчики в `clean-up`- функции
    return () => {
      document.removeEventListener("keydown", handleEscape);
      document.removeEventListener("mousedown", handleClickOutside);
    };
    // обязательно следим за `isOpen`, чтобы срабатывало только при открытии, а не при любой перерисовке компонента
  }, [isOpen, onClose, rootRef]);
}


```

Этот код помещают в отдельный файл `useClose.ts` в папке `hooks` и импортируют хук туда, где нужно.

И теперь можно использовать в любом компоненте (попапе, меню и тд), которому нужно установить эти обработчики. 

Пример того, как это использовать:
 
```js
    // создаем `ref` для нашей формы и указываем его на внешнем блоке с формой
    const formRef = useRef<HTMLElement>(null);
    
    useClose({
        isOpen: isMenuOpen,
        onClose: closeMenu,
        rootRef: formRef,
    });

```

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,49 @@
+	useOutsideClickClose({
+		isOpen,
+		rootRef: formRef,
+		onChange: () => {
```
* Когда кликаю по статье, то открывается меню боковое, а оно должно открываться только при клике на **стрелку**

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* Когда кликаю по статье, то открывается меню боковое, а оно должно открываться только при клике на **стрелку**
* `const [isMenuOpen, setIsMenuOpen] = useState(false);` нужно изначально внутри `ArticleParamsForm` объявить. Это только внутренняя логика формы. `App` не должен о ней знать

rmnkubyshkin committed:
fix: Correction with second iteration

	1. Fixed closing side form when we clicking at the article.
	   Create custom hook useClose for handling close side form
	2. Moved state isOpenMenu to ArticleParamsForm

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
